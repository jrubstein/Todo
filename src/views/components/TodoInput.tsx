import * as React from 'react'
import { sheet } from '../styles/base';

interface TodoInputProps {
    label: string,
    save: (value: string) => void
    defaultValue?: string
}

export class TodoInput extends React.Component<TodoInputProps> {
    private value:string = ''

    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
        this.updateValue = this.updateValue.bind(this)
    }

    onSubmit(event) {
        event.preventDefault()
        this.props.save(this.value)
        event.target.reset()
    }

    updateValue(e) {
        this.value = e.target.value
    }

    render() {
        return (
            <form className={sheet.classes.todoInput} onSubmit={this.onSubmit}>
                <label htmlFor="todoInput">{this.props.label}</label>
                <input className={sheet.classes.todoTextInput} type="text" id="todoInput" required onChange={this.updateValue} defaultValue={this.props.defaultValue} />
                <button className={`${sheet.classes.icon} ${sheet.classes.saveButton}`} type="submit" disabled={!!this.value.length} aria-label="save"></button>
            </form>
        )
    }
}