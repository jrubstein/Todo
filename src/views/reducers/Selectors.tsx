import { createSelector } from 'reselect'

const getItemsFilter = (state) =>  state.todos.items
const getEditingItemsFilter = (state) =>  state.todos.editingItem

export const getItemsSelector = () => {
    return createSelector(getItemsFilter, items => {
        return items
    })
}

export const getEditingItemsSelector = () => {
    return createSelector(getEditingItemsFilter, item => {
        return item
    })
}