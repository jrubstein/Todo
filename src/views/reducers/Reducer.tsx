import { initialState, TodoState } from './Store'
import { ADD_NEW_ITEM, EDIT_ITEM, SAVE_ITEM, DELETE_ITEM, TOGGLE_COMPLETE_ITEM } from '../actions/Actions'
import * as uuid from 'uuid/v4'

export const TodoReducer =  (state: TodoState = initialState.todos, action): TodoState => {
    switch (action.type) {
      case ADD_NEW_ITEM:
        const newItem = {
          id: uuid(),
          content: action.newValue,
          completed: false,
          createdAt: new Date()
        }
        return Object.assign({}, state, {
          items: [...state.items, ...[newItem]]
        })
      case EDIT_ITEM:
        return Object.assign({}, state, {
          editingItem: action.item
        })
      case DELETE_ITEM:
        return Object.assign({}, state, {
          items: state.items.filter(item => item.id !== action.id)
        })
      case SAVE_ITEM:
        const newItems = state.items.map(item => {
          if (action.id === item.id) {
            return {...item, ...{content: action.newContent}}
          }
          return item
        })

        return Object.assign({}, state, {
          items: newItems,
          editingItem: null
        })
      case TOGGLE_COMPLETE_ITEM:
        const toggledItems = state.items.map(item => {
          if (action.id === item.id) {
            return {...item, ...{completed: !item.completed}}
          }
          return item
        })
        return Object.assign({}, state, {
          items: toggledItems,
        })
      default:
        return state
    }
}