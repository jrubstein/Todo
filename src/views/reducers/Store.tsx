import { combineReducers, applyMiddleware, Store } from 'redux'
import { createStore } from 'redux'
import { TodoReducer } from './Reducer';

export interface Item  {
    id: string
    content: string
    completed: boolean
    createdAt: Date
}

export interface TodoState {
    items: Item[]
    editingItem?: Item
    showEditDialog: boolean
}

export interface State {
    todos: TodoState
}

export const initialState: State = {
    todos: {
        items: [],
        showEditDialog: false
    }
}

export const combinedReducers = combineReducers<State>({
    todos: TodoReducer
})

export const createNewStore = (init: State = initialState): Store<State> => {
    return createStore(combinedReducers , init)
    // For testing
    // !!(window || null) ? (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__() : null)
}