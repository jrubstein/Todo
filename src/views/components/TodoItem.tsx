import * as React from 'react'
import { State, Item } from '../reducers/Store'
import { connect } from 'react-redux'
import { getEditingItemsSelector } from '../reducers/Selectors'
import { editItemAction, saveItemAction, deleteItemAction, toggleCompleteItemAction } from '../actions/Actions'
import { TodoInput } from './TodoInput'
import { sheet } from '../styles/base'

const EditingItem = ({content, save}) => {
    return (
        <TodoInput save={save} label="Edit" defaultValue={content}/>
    )
}

interface TodoItemProps {
    item: Item
    editingItem?: Item
    edit?: (item: Item) => void
    save?: (id: string, content: string) => void
    delete?: (id: string) => void
    toggle?: (id: string) => void 
}

export class TodoItem extends React.Component<TodoItemProps> {

    constructor(props) {
        super(props)
        this.edit = this.edit.bind(this)
        this.save = this.save.bind(this)
        this.delete = this.delete.bind(this)
        this.toggleComplete = this.toggleComplete.bind(this)
    }

    toggleComplete() {
        this.props.toggle(this.props.item.id)
    }
    edit() {
        this.props.edit(this.props.item)
    }

    delete() {
        this.props.delete(this.props.item.id)
    }

    save(value) {
        this.props.save(this.props.item.id, value)
    }

    render() {
        const {item: {content, createdAt, id, completed}} = this.props
        const isEditing = this.props.editingItem && this.props.editingItem.id === id
        const classes = `${sheet.classes.item} ${completed ? sheet.classes.completed : ''}`
        const editButtonClasses = `${sheet.classes.icon} ${sheet.classes.editButton} ${completed ? sheet.classes.disabledEditButton : ''}`
        return (
            <div className={classes}>
                {isEditing && <EditingItem content={content} save={this.save} /> }
                {!isEditing && 
                    <React.Fragment>
                        <input type="checkbox" checked={completed} onChange={this.toggleComplete} />
                        <span className={sheet.classes.content}>{content}</span>
                        <span className={sheet.classes.date}>{createdAt.toDateString()}</span>
                        <button name="edit" className={editButtonClasses} onClick={this.edit} aria-label="edit" disabled={completed}></button>
                        <button name="delete" className={`${sheet.classes.icon} ${sheet.classes.deleteButton}`} onClick={this.delete} aria-label="delete"></button>
                    </React.Fragment>
                } 
                
            </div>
        )
    }
}

const mapStateToProps = () => {
    const getEditingItems = getEditingItemsSelector()
    return (state) => {
      return {
        editingItem: getEditingItems(state)
      }
    }
  }

  
const mapDispatchToProps = (dispatch) => ({
    edit: (item) => {
        dispatch(editItemAction(item))
    },
    delete: (id) => {
        dispatch(deleteItemAction(id))
    },
    toggle: (id) => {
        dispatch(toggleCompleteItemAction(id))
    },
    save: (id: string, newContent: string) => {
        dispatch(saveItemAction(id, newContent))
    }
})

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(TodoItem)