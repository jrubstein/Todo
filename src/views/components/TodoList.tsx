import * as React from 'react'
import { State, Item } from '../reducers/Store';
import { connect } from 'react-redux';
import TodoItem from './TodoItem';
import { getItemsSelector } from '../reducers/Selectors'
import { sheet } from '../styles/base';

class TodoList extends React.Component<{items: Item[]}> {
    render() {
        return (
            <React.Fragment>
                <h2 className={sheet.classes.numberOfItems}>{this.props.items.length} Item(s)</h2>
                {this.props.items.map(item => <TodoItem key={item.id} item={item} /> )}
            </React.Fragment>
        )
    }
}

const mapStateToProps = () => {
    const getItems = getItemsSelector()
    return (state, props) => {
      return {
        items: getItems(state)
      }
    }
  }

export default connect(
    mapStateToProps,
    null,
)(TodoList)