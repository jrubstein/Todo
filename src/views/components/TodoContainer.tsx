import * as React from 'react'
import { TodoInput } from './TodoInput'
import { State } from '../reducers/Store'
import { addNewItemAction } from '../actions/Actions'
import { connect } from 'react-redux'
import TodoList from './TodoList'
import { sheet } from '../styles/base';

class TodoContainer extends React.Component<{addNewItem?: (value: string) => void}> {
    constructor(props) {
        super(props)
        this.addNewItem = this.addNewItem.bind(this)
    }

    private addNewItem(value: string) {
        this.props.addNewItem(value)
    }

    render() {
        return (
            <div className={sheet.classes.container}>
                <h1>Todos</h1>
                <TodoList />
                <TodoInput save={this.addNewItem} label="Add todo here"/>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    addNewItem: (newValue) => {
        dispatch(addNewItemAction(newValue))
    },
})

export default connect(
    null,
    mapDispatchToProps,
)(TodoContainer)