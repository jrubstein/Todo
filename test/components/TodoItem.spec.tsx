import { create } from 'react-test-renderer'
import { createItem } from '../factories'
import * as React from 'react'
import { TodoItem } from '../../src/views/components/TodoItem'
import { shallow, configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { Provider } from 'react-redux';
import { createNewStore, initialState } from '../../src/views/reducers/Store';

configure({ adapter: new Adapter() })

describe('TodoItem test', () => {
    it('Should create valid markup', () => {
        const component = create(
            <Provider store={createNewStore(initialState)}>
                <TodoItem item={createItem()}/>
            </Provider>
        )
        expect(component).toMatchSnapshot()
    })

    it('Should create valid markup when editing', () => {
        const data = createItem()
        const component = create(
            <TodoItem item={data} editingItem={data}/>
        )
        expect(component).toMatchSnapshot()
    })

    it('Should call delete when delete button is clicked', () => {
        const data = createItem()
        const deleteFunction = jest.fn()
        const item = shallow(
            <TodoItem item={data} delete={deleteFunction}/>
        )
        item.find('button[name="delete"]').simulate('click')
        expect(deleteFunction.mock.calls.length).toEqual(1);
        expect(deleteFunction.mock.calls[0][0]).toEqual(data.id)
    })

    it('Should call edit when edit button is clicked', () => {
        const data = createItem()
        const edit = jest.fn()
        const item = shallow(
            <TodoItem item={data} edit={edit}/>
        )
        item.find('button[name="edit"]').simulate('click')
        expect(edit.mock.calls.length).toEqual(1);
        expect(edit.mock.calls[0][0]).toEqual(data)
    })

    it('Should call toggle when the checkbox is clicked', () => {
        const data = createItem()
        const toggle = jest.fn()
        const item = shallow(
            <TodoItem item={data} toggle={toggle}/>
        )
        item.find('input[type="checkbox"]').simulate('change')
        expect(toggle.mock.calls.length).toEqual(1);
        expect(toggle.mock.calls[0][0]).toEqual(data.id)
    })
})