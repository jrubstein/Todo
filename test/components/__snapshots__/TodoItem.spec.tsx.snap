// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem test Should create valid markup 1`] = `
<div
  className="item-0-0-3 "
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="content-0-0-4"
  >
    content
  </span>
  <span
    className="date-0-0-5"
  >
    Sun Jul 15 2018
  </span>
  <button
    aria-label="edit"
    className="icon-0-0-6 editButton-0-0-8 "
    disabled={false}
    name="edit"
    onClick={[Function]}
  />
  <button
    aria-label="delete"
    className="icon-0-0-6 deleteButton-0-0-7"
    name="delete"
    onClick={[Function]}
  />
</div>
`;

exports[`TodoItem test Should create valid markup when editing 1`] = `
<div
  className="item-0-0-3 "
>
  <form
    className="todoInput-0-0-11"
    onSubmit={[Function]}
  >
    <label
      htmlFor="todoInput"
    >
      Edit
    </label>
    <input
      className="todoTextInput-0-0-12"
      defaultValue="content"
      id="todoInput"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <button
      aria-label="save"
      className="icon-0-0-6 saveButton-0-0-10"
      disabled={false}
      type="submit"
    />
  </form>
</div>
`;
